bench: refactor random number generation in stats/base/dists/normal

pranav-1720 opened the pull request:
Resolves [#4981](https://github.com/stdlib-js/stdlib/issues/4981).

## Description

> What is the purpose of this pull request?

This pull request:

-   Refactors random number generation in JS benchmarks for stats/base/dists/normal.
-   Replaces randu() with uniform() for cleaner and more consistent code.
-   Moves the random number generation outside the benchmarking loops.
## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves [#4981](https://github.com/stdlib-js/stdlib/issues/4981).

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

pranav-1720 committed:
bench: refactor: Refactor random number generation in JS benchmarks for stats/base/dists/normal

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/ctor/benchmark/benchmark.js:
```diff
@@ -33,18 +34,26 @@ var Normal = require( './../lib' );
 		if ( !( dist instanceof Normal ) ) {
 			b.fail( 'should return a distribution instance' );
 		}
 	}
+
 	b.toc();
```
```suggestion
	}
	b.toc();
```

anandkaranubc replied:
This change is still needed.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/ctor/benchmark/benchmark.js:
```diff
@@ -71,6 +80,7 @@ bench( pkg+'::get:mu', function benchmark( b ) {
 			b.fail( 'should return set value' );
 		}
 	}
+
 	b.toc();
```
```suggestion
	}
	b.toc();
```

Same comment as above, and applies throughout the PR

anandkaranubc replied:
This still needs to be addressed.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/ctor/benchmark/benchmark.js:
```diff
@@ -470,11 +562,14 @@ bench( pkg+':pdf', function benchmark( b ) {
+	x = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = uniform( -3.0, 3.0 );
+	}
 	b.tic();
```
```suggestion
	}
	
	b.tic();
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/mode/benchmark/benchmark.js:
```diff
@@ -32,15 +33,22 @@ var mode = require( './../lib' );
-		mu = ( randu()*100.0 ) - 50.0;
-		sigma = ( randu()*20.0 ) + EPS;
-		y = mode( mu, sigma );
+		y = mode( mu[ i % 100 ], sigma[ i % 100 ] );
```
```suggestion
		y = mode( mu[ i % len ], sigma[ i % len ] );
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/pdf/benchmark/benchmark.js:
```diff
@@ -32,17 +33,25 @@ var pdf = require( './../lib' );
-		mu = ( randu()*100.0 ) - 50.0;
-		sigma = ( randu()*20.0 ) + EPS;
-		y = pdf( x, mu, sigma );
+		y = pdf( x[ i % 100 ], mu[ i % 100 ], sigma[ i % 100 ] );
```
Same comment.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/pdf/benchmark/benchmark.js:
```diff
@@ -66,11 +76,15 @@ bench( pkg+':factory', function benchmark( b ) {
 	for ( i = 0; i < b.iterations; i++ ) {
-		x = ( randu()*6.0 ) - 3.0;
-		y = mypdf( x );
+		y = mypdf( x[ i % 100 ] );
```
```suggestion
		y = mypdf( x[ i % len ] );
```

Applies throughout the PR.

anandkaranubc commented:
Great work, @pranav-1720! Just a few minor changes needed; otherwise, everything looks good!

And also change in title:

bench: refactor random number generation in `stats/base/dists/normal`

pranav-1720 committed:
bench: refactor random number generation in stats/base/dists/normal

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

pranav-1720 committed:
bench: refactor random number generation in stats/base/dists/normal

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

pranav-1720 commented:
@anandkaranubc hey have made the changes you suggested

pranav-1720 commented:
have made the suggested changes

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/ctor/benchmark/benchmark.js:
```diff
@@ -442,11 +529,14 @@ bench( pkg+':mgf', function benchmark( b ) {
+	x = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = uniform( 0.0, 1.0 );
+	}
 	b.tic();
 	for ( i = 0; i < b.iterations; i++ ) {
```
```suggestion
	}
	
	b.tic();
	for ( i = 0; i < b.iterations; i++ ) {
```

anandkaranubc replied:
This change is still pending

pranav-1720 committed:
bench: refactor random number generation in stats/base/dists/normal

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

pranav-1720 commented:
@anandkaranubc added the whitespace changes as well. Also is there an automated way to check these whitespace linting isues, the current linting config does not seem to catch them or perhaps i am missing something in my local setup. I get errors for trailing whitespace and some other issues but not the ones you highlighted for this PR

anandkaranubc commented:
> is there an automated way to check these whitespace linting isues

I don’t think the linter catches all the whitespace linting issues. It does catch some (there might be a reason why it is configured this way). Maybe you can look into improving it, propose an idea in the OH tomorrow, or raise an issue?

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/cdf/benchmark/benchmark.js:
```diff
@@ -32,17 +33,24 @@ var cdf = require( './../lib' );
+		mu[ i ] = uniform( -50.0, 50.0 );
+		sigma[ i ] = uniform( EPS, 20.0 );
+	}
 	b.tic();
```
```suggestion
	}
	
	b.tic();
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/ctor/benchmark/benchmark.js:
```diff
@@ -270,17 +316,24 @@ bench( pkg+':median', function benchmark( b ) {
+	len = 100;
+	x = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = uniform( EPS + 1.0, 100.0 );
```
```suggestion
		x[ i ] = uniform( 1.0 + EPS, 100.0 );
```

Not something that makes a difference, but good for consistency.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/ctor/benchmark/benchmark.js:
```diff
@@ -324,17 +384,23 @@ bench( pkg+':skewness', function benchmark( b ) {
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = uniform( EPS, 100.0 );
+	}
 	b.tic();
```
```suggestion
	}
	
	b.tic();
```

This applies throughout the PR

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/logcdf/benchmark/benchmark.js:
```diff
@@ -21,7 +21,8 @@
 var bench = require( '@stdlib/bench' );
-var randu = require( '@stdlib/random/base/randu' );
+var Float64Array = require( '@stdlib/array/float64' );
+var uniform	= require( '@stdlib/random/base/uniform' );
```
```suggestion
var uniform = require( '@stdlib/random/base/uniform' );
```

pranav-1720 replied:
fixed, how did you even find that. I dont even know how the space was replaced by tab

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/normal/logpdf/benchmark/benchmark.js:
```diff
@@ -21,7 +21,8 @@
 var bench = require( '@stdlib/bench' );
-var randu = require( '@stdlib/random/base/randu' );
+var Float64Array = require( '@stdlib/array/float64' );
+var uniform	= require( '@stdlib/random/base/uniform' );
```
Same comment

anandkaranubc reviewed (changes requested):
Still some changes to be made. Rest, the PR looks good. Thanks for making those changes!

pranav-1720 committed:
bench: refactor random number generation in stats/base/dists/normal

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

pranav-1720 committed:
bench: refactor random number generation in stats/base/dists/normal

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

pranav-1720 commented:
@anandkaranubc made the changes, hopefully everything should be in order now. Cheers!

Planeshifter commented:
We have moved to using `var uniform = require( '@stdlib/random/array/uniform' );` instead of `var uniform = require( '@stdlib/random/base/uniform' );` to directly draw arrays filled with numbers from the desired distribution. Since this PR follows our older convention and now has merge conflicts, I am going to close it. Thanks for your contribution either way, it is appreciated!
